File: src/clients/desktop_app/ami_daemon/routers/integrations.py
"""
Integrations API Router

Manages cloud service integrations (Gmail, Google Drive, Calendar, Notion).
Based on Eigent's integration management patterns.

Features:
- OAuth flow handling for Google services
- Token-based authentication for Notion
- Integration status tracking
- Credential storage (secure)
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
import logging
import os
import json
from pathlib import Path

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/integrations", tags=["Integrations"])


# ============== Data Storage ==============

def get_integrations_file() -> Path:
    """Get path to integrations storage file."""
    ami_dir = Path.home() / ".ami"
    ami_dir.mkdir(exist_ok=True)
    return ami_dir / "integrations.json"


def load_integrations() -> Dict[str, Any]:
    """Load integrations from storage."""
    file_path = get_integrations_file()
    if file_path.exists():
        try:
            with open(file_path, 'r') as f:
                return json.load(f)
        except Exception as e:
            logger.error(f"Failed to load integrations: {e}")
    return {"installed": [], "configs": {}}


def save_integrations(data: Dict[str, Any]) -> None:
    """Save integrations to storage."""
    file_path = get_integrations_file()
    try:
        with open(file_path, 'w') as f:
            json.dump(data, f, indent=2)
    except Exception as e:
        logger.error(f"Failed to save integrations: {e}")


# ============== Request/Response Models ==============

class IntegrationConfigRequest(BaseModel):
    """Configuration for token-based integrations."""
    api_key: Optional[str] = Field(None, description="API key (e.g., Notion)")


class IntegrationListResponse(BaseModel):
    """List of installed integrations."""
    installed: List[str]
    available: List[str] = [
        "gmail", "google_drive", "google_calendar", "notion"
    ]


class OAuthStatusResponse(BaseModel):
    """OAuth flow status."""
    completed: bool = False
    failed: bool = False
    error: Optional[str] = None


# ============== OAuth State Tracking ==============

# In-memory OAuth state (would be Redis in production)
_oauth_states: Dict[str, Dict[str, Any]] = {}


# ============== Endpoints ==============

@router.get("/list", response_model=IntegrationListResponse)
async def list_integrations():
    """
    List installed integrations.

    Returns list of integration IDs that have been configured.
    """
    data = load_integrations()
    return IntegrationListResponse(
        installed=data.get("installed", []),
        available=["gmail", "google_drive", "google_calendar", "notion"]
    )


@router.get("/oauth-status/{integration_id}", response_model=OAuthStatusResponse)
async def get_oauth_status(integration_id: str):
    """
    Get OAuth flow status for an integration.

    Frontend polls this endpoint after initiating OAuth to check completion.
    """
    state = _oauth_states.get(integration_id)
    if not state:
        # No active OAuth flow - check if already installed
        data = load_integrations()
        if integration_id in data.get("installed", []):
            return OAuthStatusResponse(completed=True)
        return OAuthStatusResponse(completed=False, failed=False)

    return OAuthStatusResponse(
        completed=state.get("completed", False),
        failed=state.get("failed", False),
        error=state.get("error")
    )


@router.post("/oauth-callback/{integration_id}")
async def oauth_callback(integration_id: str, code: str, state: Optional[str] = None):
    """
    Handle OAuth callback.

    This endpoint receives the authorization code after user consents.
    Exchanges code for tokens and stores credentials.
    """
    try:
        # For Google integrations, exchange code for tokens
        if integration_id in ("gmail", "google_drive", "google_calendar"):
            # In production, this would:
            # 1. Exchange auth code for access/refresh tokens
            # 2. Store tokens securely
            # 3. Mark integration as installed

            # For now, mark as completed (frontend handles OAuth via Tauri)
            _oauth_states[integration_id] = {"completed": True}

            # Update storage
            data = load_integrations()
            if integration_id not in data["installed"]:
                data["installed"].append(integration_id)
                save_integrations(data)

            logger.info(f"OAuth completed for {integration_id}")
            return {"success": True, "message": "OAuth completed"}

        else:
            raise HTTPException(status_code=400, detail="Invalid integration for OAuth")

    except Exception as e:
        logger.error(f"OAuth callback error: {e}")
        _oauth_states[integration_id] = {"failed": True, "error": str(e)}
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/configure/{integration_id}")
async def configure_integration(integration_id: str, config: IntegrationConfigRequest):
    """
    Configure a token-based integration.

    For integrations like Notion that use API keys instead of OAuth.
    """
    try:
        # Validate the configuration
        if integration_id == "notion":
            if not config.api_key or not config.api_key.startswith("secret_"):
                raise HTTPException(
                    status_code=400,
                    detail="Invalid Notion API key. Must start with 'secret_'"
                )

            # Store configuration
            data = load_integrations()
            if "configs" not in data:
                data["configs"] = {}
            data["configs"][integration_id] = {
                "api_key": config.api_key,  # In production, encrypt this
                "configured_at": str(os.popen('date').read().strip())
            }

            if integration_id not in data["installed"]:
                data["installed"].append(integration_id)

            save_integrations(data)
            logger.info(f"Configured {integration_id}")

            return {"success": True, "message": f"{integration_id} configured successfully"}

        else:
            raise HTTPException(status_code=400, detail=f"Unknown integration: {integration_id}")

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Configure error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/uninstall/{integration_id}")
async def uninstall_integration(integration_id: str):
    """
    Uninstall an integration.

    Removes credentials and marks as uninstalled.
    """
    try:
        data = load_integrations()

        # Remove from installed list
        if integration_id in data.get("installed", []):
            data["installed"].remove(integration_id)

        # Remove configuration
        if integration_id in data.get("configs", {}):
            del data["configs"][integration_id]

        save_integrations(data)

        # Clear OAuth state
        if integration_id in _oauth_states:
            del _oauth_states[integration_id]

        logger.info(f"Uninstalled {integration_id}")
        return {"success": True, "message": f"{integration_id} uninstalled"}

    except Exception as e:
        logger.error(f"Uninstall error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/config/{integration_id}")
async def get_integration_config(integration_id: str):
    """
    Get configuration for an integration (without secrets).

    Returns metadata about configuration without exposing API keys.
    """
    data = load_integrations()

    if integration_id not in data.get("installed", []):
        raise HTTPException(status_code=404, detail="Integration not installed")

    config = data.get("configs", {}).get(integration_id, {})

    # Return config without sensitive data
    return {
        "integration_id": integration_id,
        "installed": True,
        "configured_at": config.get("configured_at"),
        "has_api_key": bool(config.get("api_key")),
    }

File: src/clients/desktop_app/daemon-ts/src/routes/integrations.ts
/**
 * Integrations Routes — OAuth & API key management for third-party services.
 *
 * Ported from integrations.py.
 *
 * GET  /api/v1/integrations/list                 — List installed/available integrations
 * GET  /api/v1/integrations/oauth-status/:id      — Check OAuth flow status
 * POST /api/v1/integrations/oauth-callback/:id    — Handle OAuth callback
 * POST /api/v1/integrations/configure/:id         — Configure API-key integration
 * POST /api/v1/integrations/uninstall/:id         — Remove integration
 * GET  /api/v1/integrations/config/:id            — Get integration config (no secrets)
 */

import { Router, type Request, type Response } from "express";
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { createLogger } from "../utils/logging.js";

const logger = createLogger("integrations");

export const integrationsRouter = Router();

// ===== Storage =====

const AMI_DIR = join(homedir(), ".ami");
const INTEGRATIONS_FILE = join(AMI_DIR, "integrations.json");

const AVAILABLE_INTEGRATIONS = ["gmail", "google_drive", "google_calendar", "notion"];

interface IntegrationData {
  installed: string[];
  configs: Record<string, { api_key?: string; configured_at?: string }>;
}

function loadIntegrations(): IntegrationData {
  try {
    if (existsSync(INTEGRATIONS_FILE)) {
      return JSON.parse(readFileSync(INTEGRATIONS_FILE, "utf-8"));
    }
  } catch {
    // corrupted file
  }
  return { installed: [], configs: {} };
}

function saveIntegrations(data: IntegrationData): void {
  mkdirSync(AMI_DIR, { recursive: true });
  writeFileSync(INTEGRATIONS_FILE, JSON.stringify(data, null, 2), "utf-8");
}

// ===== In-memory OAuth state =====

const oauthStates = new Map<string, { completed: boolean; failed: boolean; error?: string }>();

// ===== GET /list =====

integrationsRouter.get("/list", (_req: Request, res: Response) => {
  const data = loadIntegrations();
  res.json({
    installed: data.installed,
    available: AVAILABLE_INTEGRATIONS,
  });
});

// ===== GET /oauth-status/:integrationId =====

integrationsRouter.get("/oauth-status/:integrationId", (req: Request, res: Response) => {
  const { integrationId } = req.params;
  const state = oauthStates.get(integrationId);

  if (state) {
    res.json(state);
    return;
  }

  // Check if already installed
  const data = loadIntegrations();
  if (data.installed.includes(integrationId)) {
    res.json({ completed: true, failed: false, error: null });
    return;
  }

  res.json({ completed: false, failed: false, error: null });
});

// ===== POST /oauth-callback/:integrationId =====

integrationsRouter.post("/oauth-callback/:integrationId", (req: Request, res: Response) => {
  const { integrationId } = req.params;
  const code = req.query.code as string | undefined;

  const googleServices = ["gmail", "google_drive", "google_calendar"];
  if (!googleServices.includes(integrationId)) {
    res.status(400).json({ error: `Invalid integration: ${integrationId}` });
    return;
  }

  try {
    oauthStates.set(integrationId, { completed: true, failed: false });

    const data = loadIntegrations();
    if (!data.installed.includes(integrationId)) {
      data.installed.push(integrationId);
    }
    saveIntegrations(data);

    logger.info({ integrationId }, "OAuth completed");
    res.json({ success: true, message: "OAuth completed" });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    oauthStates.set(integrationId, { completed: false, failed: true, error: msg });
    res.status(500).json({ error: msg });
  }
});

// ===== POST /configure/:integrationId =====

integrationsRouter.post("/configure/:integrationId", (req: Request, res: Response) => {
  const { integrationId } = req.params;
  const { api_key } = req.body;

  if (integrationId !== "notion") {
    res.status(400).json({ error: `Configuration not supported for: ${integrationId}` });
    return;
  }

  if (!api_key || typeof api_key !== "string" || !api_key.startsWith("secret_")) {
    res.status(400).json({ error: "Invalid API key format (must start with 'secret_')" });
    return;
  }

  try {
    const data = loadIntegrations();
    data.configs[integrationId] = {
      api_key,
      configured_at: new Date().toISOString(),
    };
    if (!data.installed.includes(integrationId)) {
      data.installed.push(integrationId);
    }
    saveIntegrations(data);

    logger.info({ integrationId }, "Integration configured");
    res.json({ success: true, message: `${integrationId} configured successfully` });
  } catch (err) {
    res.status(500).json({ error: `Configuration failed: ${err}` });
  }
});

// ===== POST /uninstall/:integrationId =====

integrationsRouter.post("/uninstall/:integrationId", (req: Request, res: Response) => {
  const { integrationId } = req.params;

  try {
    const data = loadIntegrations();
    data.installed = data.installed.filter((id) => id !== integrationId);
    delete data.configs[integrationId];
    saveIntegrations(data);

    oauthStates.delete(integrationId);

    logger.info({ integrationId }, "Integration uninstalled");
    res.json({ success: true, message: `${integrationId} uninstalled` });
  } catch (err) {
    res.status(500).json({ error: `Uninstall failed: ${err}` });
  }
});

// ===== GET /config/:integrationId =====

integrationsRouter.get("/config/:integrationId", (req: Request, res: Response) => {
  const { integrationId } = req.params;
  const data = loadIntegrations();

  if (!data.installed.includes(integrationId)) {
    res.status(404).json({ error: `Integration not installed: ${integrationId}` });
    return;
  }

  const config = data.configs[integrationId];
  res.json({
    integration_id: integrationId,
    installed: true,
    configured_at: config?.configured_at ?? null,
    has_api_key: !!config?.api_key,
  });
});
